src/buffer/out/search.cpp

```diff
@@ -20,324 +17,83 @@ using namespace Microsoft::Console::Types;
 // - textBuffer - The screen text buffer to search through (the "haystack")
 // - renderData - The IRenderData type reference, it is for providing selection methods
 // - str - The search term you want to find (the "needle")
-// - direction - The direction to search (upward or downward)
-// - sensitivity - Whether or not you care about case
-Search::Search(Microsoft::Console::Render::IRenderData& renderData,
-               const std::wstring_view str,
-               const Direction direction,
-               const Sensitivity sensitivity) :
-    _direction(direction),
-    _sensitivity(sensitivity),
-    _needle(s_CreateNeedleFromString(str)),
-    _renderData(renderData),
-    _coordAnchor(s_GetInitialAnchor(renderData, direction))
+// - reverse - True when searching backward/upwards in the buffer
+// - caseInsensitive - As the name indicates: case insensitivity
+Search::Search(Microsoft::Console::Render::IRenderData& renderData, const std::wstring_view str, bool reverse, bool caseInsensitive) :
+    _renderData{ &renderData },
+    _step{ reverse ? -1 : 1 }
 {
-    _coordNext = _coordAnchor;
-}
+    const auto& textBuffer = _renderData->GetTextBuffer();
 
-// Routine Description:
-// - Constructs a Search object.
-// - Make a Search object then call .FindNext() to locate items.
-// - Once you've found something, you can perform actions like .Select() or .Color()
-// Arguments:
-// - textBuffer - The screen text buffer to search through (the "haystack")
-// - renderData - The IRenderData type reference, it is for providing selection methods
-// - str - The search term you want to find (the "needle")
-// - direction - The direction to search (upward or downward)
-// - sensitivity - Whether or not you care about case
-// - anchor - starting search location in screenInfo
-Search::Search(Microsoft::Console::Render::IRenderData& renderData,
-               const std::wstring_view str,
-               const Direction direction,
-               const Sensitivity sensitivity,
-               const til::point anchor) :
-    _direction(direction),
-    _sensitivity(sensitivity),
-    _needle(s_CreateNeedleFromString(str)),
-    _coordAnchor(anchor),
-    _renderData(renderData)
-{
-    _coordNext = _coordAnchor;
-}
+    _results = textBuffer.SearchText(str, caseInsensitive);
+    _mutationCount = textBuffer.GetMutationCount();
 

```

Comment:
We could only assert that the lock is being held, but I think that this might be better as a separate PR that does this for all the parts that implicitly rely on the console lock being held. The mutation count could be made atomic, but this won't realistically improve anything, because you need to hold the lock anyways to initiate any kind of search (either to do a new search or to select stuff in the buffer).